src/sqlfluff/core/parser/segments/base.py

```diff
@@ -1065,33 +1080,40 @@ def recursive_crawl(
                         no_recursive_seg_type=no_recursive_seg_type,
                     )
 
-    def path_to(self, other):
+    def path_to(self, other) -> List[PathStep]:
         """Given a segment which is assumed within self, get the intermediate segments.
 
         Returns:
-            :obj:`list` of segments, including the segment we're looking for.
-            None if not found.
-
+            :obj:`list` of :obj:`PathStep`, not including the segment we're looking
+                for. If `other` is not found, then empty list. This includes if
+                called on self.
         """
-        # Return self if we've found the segment.
+        # Return empty if we've found the segment.
         if self is other:
-            return [self]
+            return []  # pragma: no cover
 
         # Are we in the right ballpark?
         # NB: Comparisons have a higher precedence than `not`.
         if not self.get_start_loc() <= other.get_start_loc() <= self.get_end_loc():
-            return None
+            return []
```

Comment:
I have sometimes wondered if this result should be indicated a different way. It could be an error in some situations.

Also, does this result become ambiguous now that the last step is removed? It could be interpreted in two ways:
* There is no path to `other` from `self`.
* `other` is an immediate child of `self`.

Reply:
So my interpretation of this is that it's ok for `[]` to be the only result and not require an additional `None` result. The reason is that `[]` should always be interpreted as `other` _is not a descendent of_ `self`. That interpretation would include `other is self`, because something is not a descendent of itself.

I'll update the docstring to be clearer on that - but does that make sense?

Reply:
I've updated the docstring with the explanation:

```rst
The result of this should be interpreted as *the path from `self` to `other`*.
If the return value is `[]` (an empty list), that implies there is no path
from `self` to `other`. This would include the case where the two are the same
segment, as there is no path from a segment to itself.
```

Reply:
Oh, I see -- if `other` is an immediate child, it returns a path containing `self`. The code is definitely good. In mathematical terms, the path we're returning would be described as a ["half open interval"](https://mathresearch.utsa.edu/wiki/index.php?title=Interval_Notation#:~:text=A%20half%2Dopen%20interval%20includes,0%20and%20less%20than%201.), which we could mention in the docs as `self, other)`. I think it would be cool add that additional info, because it's formal and precise about what's being returned. Sometimes words are not enough. 😛

Reply:
@barrywhart - I've added some better tests to spell it out better and a not in the docstring. Reckon that's better?

Reply:
🎉♥️